graphql_compiler/compiler/emit_sql.py

```diff
@@ -929,11 +963,21 @@ def emit_code_from_ir(sql_schema_info, ir):
         SQLAlchemy Query
     """
     state = CompilationState(sql_schema_info, ir)
-    for block in _traverse_and_validate_blocks(ir):
+    for index, block in enumerate(_traverse_and_validate_blocks(ir)):
         if isinstance(block, blocks.QueryRoot):
             pass
         elif isinstance(block, blocks.MarkLocation):
-            state.mark_location()
+            # If we are in a fold and the next block either backtracks
+            # or unfolds, then we have traversed to the deepest point
+            # of the fold and need to set the output columns
+            backtracks_next = isinstance(ir.ir_blocks[index + 1], blocks.Backtrack)
+            unfolds_next = isinstance(ir.ir_blocks[index + 1], blocks.Unfold)
+            if state.is_in_fold() and (unfolds_next or backtracks_next):
+                is_last_fold_block = True
+            else:
+                is_last_fold_block = False
```

Comment:
This needs to be untangled a bit. Let me walk you through my experience as a reader:
1. We see if we're in the deepest point in the fold
2. We rephrase `deepest` as `last`, adding confusion. The meaning of deepest was documented inside the comment, but what last means is left for me to guess.
3. We pass `is_last_fold_block` into `mark_location`, where this happens:
```
        if is_last_fold_block:
            # Pull out the output columns on the last block inside the fold.
            self._current_fold.visit_output_vertex(...
```
I'm supposed to understand that the output columns need to be pulled on the last block. With no memory of what "pulling columns" menas or what the "last block is", this is a large jump in reasoning, and the only comment explaining this is in a totally different place. Furthermore, the function is not called `pull_output_columns` as this comment leads me to believe, but `visit_output_vertex`, which judging by the name may entail many more things .
4. We look inside the `visit_output_vertex` function documentation and see... `Update output columns when visiting`, finally unraveling the mystery of the plumbing. Now I can finally get into trying to understand the content of this code.

This code seems understandable when you work on it, but give it 2 months and you'll have no idea why this is happening. To tidy up, you have to have a higher respect for function specifications. Just follow these 3 rules:
1. For each function, the name and spec fully define the function
2. The implementation of the function implements the spec and nothing else
3. The usage of the function uses the spec and does not rely on anything else
4. The tests of the function test that the implementation implements the spec and nothing else

Reply:
Hi, good comments, I agree completely. Let me know what you think of the code I just pushed. It's possible also that changing the name or documentation of `mark_location` is necessary, but I'm not sure if we need to jump to that if the current docs suffice.